pkg/virtualgateway/resource_manager.go

```diff
@@ -0,0 +1,270 @@
+		if err != nil {
+			return err
+		}
+	} else {
+		sdkVG, err = m.updateSDKVirtualGateway(ctx, sdkVG, ms, vg)
+		if err != nil {
+			return err
+		}
+	}
+
+	return m.updateCRDVirtualGateway(ctx, vg, sdkVG)
+}
+
+func (m *defaultResourceManager) Cleanup(ctx context.Context, vg *appmesh.VirtualGateway) error {
+	ms, err := m.findMeshDependency(ctx, vg)
+	if err != nil {
+		return err
+	}
+	sdkVG, err := m.findSDKVirtualGateway(ctx, ms, vg)
+	if sdkVG == nil {
+		return nil
+	}
+
+	return m.deleteSDKVirtualGateway(ctx, sdkVG, ms, vg)
+}
+
+// findMeshDependency find the Mesh dependency for this virtualGateway.
+func (m *defaultResourceManager) findMeshDependency(ctx context.Context, vg *appmesh.VirtualGateway) (*appmesh.Mesh, error) {
+	if vg.Spec.MeshRef == nil {
+		return nil, errors.Errorf("meshRef shouldn't be nil, please check webhook setup")
+	}
+	ms, err := m.referencesResolver.ResolveMeshReference(ctx, *vg.Spec.MeshRef)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to resolve meshRef")
+	}
+	return ms, nil
+}
+
+// validateMeshDependencies validate the Mesh dependency for this virtualGateway.
+func (m *defaultResourceManager) validateMeshDependencies(ctx context.Context, ms *appmesh.Mesh) error {
+	if !mesh.IsMeshActive(ms) {
+		return runtime.NewRequeueError(errors.New("mesh is not active yet"))
+	}
+	return nil
+}
+
+func (m *defaultResourceManager) findSDKVirtualGateway(ctx context.Context, ms *appmesh.Mesh, vg *appmesh.VirtualGateway) (*appmeshsdk.VirtualGatewayData, error) {
+	resp, err := m.appMeshSDK.DescribeVirtualGatewayWithContext(ctx, &appmeshsdk.DescribeVirtualGatewayInput{
+		MeshName:           ms.Spec.AWSName,
+		MeshOwner:          ms.Spec.MeshOwner,
+		VirtualGatewayName: vg.Spec.AWSName,
+	})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "NotFoundException" {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return resp.VirtualGateway, nil
+}
+
+func (m *defaultResourceManager) createSDKVirtualGateway(ctx context.Context, ms *appmesh.Mesh, vg *appmesh.VirtualGateway) (*appmeshsdk.VirtualGatewayData, error) {
+	sdkVGSpec, err := m.buildSDKVirtualGatewaySpec(ctx, vg)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := m.appMeshSDK.CreateVirtualGatewayWithContext(ctx, &appmeshsdk.CreateVirtualGatewayInput{
+		MeshName:           ms.Spec.AWSName,
+		MeshOwner:          ms.Spec.MeshOwner,
+		Spec:               sdkVGSpec,
+		Tags:               nil,
+		VirtualGatewayName: vg.Spec.AWSName,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return resp.VirtualGateway, nil
+}
+
+func (m *defaultResourceManager) updateSDKVirtualGateway(ctx context.Context, sdkVG *appmeshsdk.VirtualGatewayData, ms *appmesh.Mesh, vg *appmesh.VirtualGateway) (*appmeshsdk.VirtualGatewayData, error) {
+	actualSDKVGSpec := sdkVG.Spec
+	desiredSDKVGSpec, err := m.buildSDKVirtualGatewaySpec(ctx, vg)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := cmpopts.EquateEmpty()
```

Comment:
see https://github.com/aws/aws-app-mesh-controller-for-k8s/pull/218#discussion_r433386446  :D

Reply:
You're right, port is set by the server if left unset from the controller side. I'll fix the comp. Good catch 👍